### PR TITLE
Refactor Metal texture uploads

### PR DESCRIPTION
**Pull request by bejado**
Previously, Metal was never performing any texture conversions during upload. For example, `iblprefilter` loads `FLOAT` image data into the `R11F_G11_B10F` texture format. This PR fixes the issue, refactors the upload pipeline, and adds a new backend test for coverage.

### CODE REVIEW

**Comment by pixelflinger** [comment]
Is this going in? :-)

**Comment by bejado** [comment]
@pixelflinger Yes, I still need to fix sRGB texture uploads, then it will be ready for review.

**Review by pixelflinger — APPROVED** [review]

File: filament/backend/src/BackendUtils.cpp [hunk 1]
```diff
@@ -326,6 +328,55 @@ size_t getBlockHeight(TextureFormat format) noexcept {
     }
 }
 
+bool reshape(const PixelBufferDescriptor& data, PixelBufferDescriptor& reshaped) {
+    if (data.format != PixelDataFormat::RGB && data.format != PixelDataFormat::RGB_INTEGER) {
```
Comment (pixelflinger) [inline filament/backend/src/BackendUtils.cpp:332]: can we get a comment for this test, I'm not understanding it.

File: filament/backend/src/metal/MetalHandles.h [hunk 2]
```diff
@@ -164,6 +163,19 @@ struct MetalProgram : public HwProgram {
     bool isValid = false;
 };
 
+struct PixelBufferShape {
+    size_t bytesPerPixel;
+    size_t bytesPerRow;
+    size_t bytesPerSlice;
+    size_t totalBytes;
```
Comment (pixelflinger) [inline filament/backend/src/metal/MetalHandles.h:170]: do all these need to be `size_t` as opposed to `uint32_t`?

File: filament/backend/src/metal/MetalHandles.mm [hunk 3]
```diff
@@ -537,162 +521,242 @@ static MTLPixelFormat decidePixelFormat(id<MTLDevice> device, TextureFormat form
...
+    const size_t stagingBufferSize = shape.totalBytes;
+    const bool largeUpload = stagingBufferSize > deviceMaxBufferLength;
+
+    ASSERT_PRECONDITION(!nonBlittableTexture || !conversionNecessary,
+            "SAMPLER_2D_ARRAY, SAMPLER_3D, and SAMPLER_CUBEMAP texture uploads"
+            "do not support format conversions.");
+
+    ASSERT_PRECONDITION(!nonBlittableTexture || !largeUpload,
+            "SAMPLER_2D_ARRAY, SAMPLER_3D, and SAMPLER_CUBEMAP texture uploads"
+            "have a max size of %d bytes.", deviceMaxBufferLength);
```
Comment (pixelflinger) [inline filament/backend/src/metal/MetalHandles.mm:674]: Is this a "todo" or just something we don't support in filament? If it's the later, we should probably have the ASSERT_PRECONDITION on the user facing API, and just use an assert() here.

**Review by pixelflinger — APPROVED** [review]
I'm not 100% sure I understand all the steps needed. If I'm getting this right, we use the DataReshaper (CPU) only when the source format is RGB (but all other cases, R, RG, RGBA bypass the DataReshaper), then the actual texture upload is done with the GPU? Correct?

What I'm not clear about is if the srd->dst conversions we handle are the same and exactly the same than GLES' (i.e. no more, no less)?

**Comment by bejado** [comment]
> If I'm getting this right, we use the DataReshaper (CPU) only when the source format is RGB (but all other cases, R, RG, RGBA bypass the DataReshaper), then the actual texture upload is done with the GPU? Correct?

Correct. Since Metal does not support 3-component textures, we have to reshape the data on the CPU first. Then we use the GPU to upload to the actual texture, which also performs any necessary conversions.

> What I'm not clear about is if the srd->dst conversions we handle are the same and exactly the same than GLES' (i.e. no more, no less)?

I made sure to support at least what GLES supports. In practice however, the Metal backend should support additional conversions, for instance, `RGB32F` to `RGBA16F`. I don't think GLES explicitly supports that one.

File: filament/backend/src/metal/MetalHandles.mm (hunk 3 above)
Comment (bejado) [reply filament/backend/src/metal/MetalHandles.mm:674]: This is a limitation of the Metal backend at the moment, so more of a "todo". I'll add a TODO comment.

File: filament/backend/src/metal/MetalHandles.h (hunk 2 above)
Comment (bejado) [reply filament/backend/src/metal/MetalHandles.h:170]: `uint32_t` should be fine.

**Pull request merged**